### PR TITLE
Update "Usage with React"

### PR DESCRIPTION
**Pull request by gaearon**
This builds on top of https://github.com/rackt/redux/pull/1186 with some changes I wanted to make to the wording and structure. "Usage with React" looks done to me (although improvements are welcome!)

Right now I really want somebody to contribute a few things here:

* @pedrottimark **(Done!)** <s>Update https://github.com/rackt/redux/blob/___react-docs-update/docs/basics/ExampleTodoList.md to match the source code from the new "Usage with React"</s>
* @jakegardner <s>Update https://github.com/rackt/redux/blob/___react-docs-update/docs/recipes/ComputingDerivedData.md to match the new example (should be just a few tweaks, e.g. it should feature VisibleTodoList rather than App component)</s>
* @gaearon <s>Update https://github.com/rackt/redux/tree/___react-docs-update/examples/todos-with-undo to also match that source code (except for the undo part, obviously)</s>
* @gaearon <s>Update https://github.com/rackt/redux/blob/___react-docs-update/docs/recipes/ImplementingUndoHistory.md to match the new `todos-with-undo` example</s>

Help greatly appreciated because I'm busy but I really really want to ship this.
Please post in this thread if you'd like to take some part of this work.

### CODE REVIEW

**Comment by gaearon** [comment]
cc @omnidan who owns `todos-with-undo`, maybe you've got some time to help?

**Comment by pedrottimark** [comment]
I will go for the first bullet point: basics/ExampleTodoList.md

**Comment by gaearon** [comment]
To make it super clear, the new "Usage with React" is here:
https://github.com/rackt/redux/blob/___react-docs-update/docs/basics/UsageWithReact.md

**Comment by urbanvikingr** [comment]
@gaearon @pedrottimark Perhaps this can be helpful. It is the source code used for the update to Usage with React. I didn't include tests, otherwise the code is complete.

**Comment by pedrottimark** [comment]
@urbanvikingr Do you think it would be more efficient and less risk of error for you to do it? No offense taken, if you have time and want to.

**Comment by gaearon** [comment]
Note there were mixed code styles in previous PR. I tried to amend those to match our code style enforced by the linter.

**Comment by pedrottimark** [comment]
@gaearon So the goal is on ___react-docs-update branch: copy each chunk of code from UsageWithReact.md paste it where it goes in the order of ExampleTodoList.md but leave all other code there which is not in the intersection unchanged (for example, Action Creators and Constants).

**Comment by pedrottimark** [comment]
@gaearon Hopefully the Pooh Bear brain of a hybrid writer-designer-developer doesn’t drive you crazy.

1. In Usage, the Presentational Components have two different orders.
It might help smooth things out if the order under Implementing Components changes to become consistent with the order previously established under Designing Component Hierarchy: TodoList, Todo, Link, Footer, App.
If you agree, should I include that in the pull request?

2. What do you think if the major and minor order of source for the components in the lower half of Examples becomes parallel with Usage?
That is, Presentational Components, Container Components, Other Components
Specific minor order depends on answer to preceding question.

**Comment by gaearon** [comment]
>So the goal is on ___react-docs-update branch: copy each chunk of code from UsageWithReact.md paste it where it goes in the order of ExampleTodoList.md but leave all other code there which is not in the intersection unchanged (for example, Action Creators and Constants).

For the rest of the code, use code from https://github.com/urbanvikingr/todo/. It's based on my Egghead tutorials.

>It might help smooth things out if the order under Implementing Components changes to become consistent with the order previously established under Designing Component Hierarchy: TodoList, Todo, Link, Footer, App.

I agree order should be consistent. We should strive to maintain “first declared, then used” order. So `Todo` should be before `TodoList`.

**Comment by gaearon** [comment]
>What do you think if the major and minor order of source for the components in the lower half of Examples becomes parallel with Usage?

Yes, order everywhere should be consistent. Thanks for helping!

**Comment by pedrottimark** [comment]
@gaearon Thank you as always for your patience to explain. Super, I will use “first declared, then used” order as the guiding principle when there is an inconsistency. Am keeping notes as other little things come up to ask as a batch tomorrow.

**Comment by gaearon** [comment]
:+1:

**Comment by gaearon** [comment]
@pedrottimark 

I gave you the commit bit so please feel free to edit the branch right as part of this PR. I'd like if you could flush your temporary progress so we have some shared bits others could use for the rest of the bullet points.

**Comment by urbanvikingr** [comment]
@pedrottimark I can update source code for ToDo this weekend.

**Comment by jakegardner** [comment]
I'll take the second bullet: recipes/ComputingDerivedData.md

**Comment by gaearon** [comment]
@jakegardner Thank you!

**Comment by pedrottimark** [comment]
@gaearon Three questions about your house style for source code:

1. Quotation marks: Have previously been single. Videos use double. So now Usage has single in import statements and double in the updated code. Okay that way?

2. Constants for actions and filters: Videos and updated code in project from which I pasted use string literals instead of constants. Okay?

3. propTypes for component functions: So far I kept the existing propTypes and added it for one new function. Keep or remove?

**Comment by gaearon** [comment]
For style, check what our linter settings allow. You can paste code into examples and `npm run lint` in the root to verify. From what I remember our settings prefer single quotes in JS but double quotes in JSX.

I don't mind string liberals in this example.

We should always use propTypes. If they are missing we should add them.

**Comment by pedrottimark** [comment]
@gaearon You remember correctly about the quote marks. Am finished unless I find anything incorrect or inconsistent when I proofread the two files in Markdown viewer tomorrow with fresh eyes.

@urbanvikingr Decided to go ahead. Thank you for putting the ball in play by updating according to the videos. The smaller chunks make these pages even easier to understand than before.

**Comment by gaearon** [comment]
@pedrottimark Thank you very much. Can we check off the first item from the todo list?

**Comment by gaearon** [comment]
(I meant the todo list in the first post in this thread :-)

**Comment by pedrottimark** [comment]
@gaearon Yes!

**Comment by gaearon** [comment]
:+1: 

@jakegardner Still up for the second point? We've got the sample code you can base the article update upon in https://github.com/rackt/redux/blob/___react-docs-update/docs/basics/ExampleTodoList.md now. You could just concentrate on these functions:

```js
const getVisibleTodos = (todos, filter) => {
  switch (filter) {
    case 'SHOW_ALL':
      return todos
    case 'SHOW_COMPLETED':
      return todos.filter(t => t.completed)
    case 'SHOW_ACTIVE':
      return todos.filter(t => !t.completed)
  }
}

const mapStateToProps = (state) => {
  return {
    todos: getVisibleTodos(state.todos, state.visibilityFilter)
  }
}
```

and show how to re-implement them with Reselect efficiently.

**Comment by jakegardner** [comment]
Sure, working on it now.

**Comment by urbanvikingr** [comment]
@pedrottimark Thank you.

@gaearon I will update todos-with-undo example. It is very similar to my update of the todo example.

**Comment by gaearon** [comment]
@jakegardner @urbanvikingr Awesome, thx!

**Comment by jakegardner** [comment]
Ok, pretty much ready for review, just one part I'm not sure about.

Should the visibleTodosSelector still be added to the connect function, and assign todos to visibleTodos -OR- just use visibleTodosSelector directly in the todos assignment?

**Comment by gaearon** [comment]
Can you show how code is different for these cases?

File: docs/basics/ExampleTodoList.md [hunk 1]
```diff
@@ -9,116 +9,85 @@ This is the complete source code of the tiny todo app we built during the [basic
...
 import React from 'react'
 import { render } from 'react-dom'
-import { createStore } from 'redux'
 import { Provider } from 'react-redux'
-import App from './containers/App'
+import { createStore } from 'redux'
 import todoApp from './reducers'
+import App from './components/App'
 
 let store = createStore(todoApp)
```
Comment (rockingskier) [inline docs/basics/ExampleTodoList.md]: Should this be `todoStore`?
https://github.com/rackt/redux/pull/1285/files#diff-e9a37286e0b5c53a133d6291766a478fR20

Comment (gaearon) [inline docs/basics/ExampleTodoList.md]: I'd rather change it below. It should be just `store` to avoid implying that we have many stores.

**Comment by jakegardner** [comment]
Using connect:
```js
import { visibleTodosSelector } from '../selectors/todoSelectors'

const mapStateToProps = (state) => {
  return {
    todos: state.visibleTodos // ?
  }
}
...
const VisibleTodoList = connect(
  visibleTodosSelector,
  mapStateToProps,
  mapDispatchToProps
)(TodoList)
```

Without:
```js
...
import { visibleTodosSelector } from '../selectors/todoSelectors'

const mapStateToProps = (state) => {
  return {
    todos: visibleTodosSelector.visibleTodos // ?
  }
}
...
const VisibleTodoList = connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoList)
```

Neither of these seem exactly right to me. If we add visibleTodosSelector to connect, that should expose visibleTodos and visibilityFilter to props, right? Then would just need to update TodoList to accept visibleTodos, but not sure how visibilityFilter then maps to FilterLink.

You can see my progress so far at [https://github.com/rackt/redux/commit/5ccf18b0e1055e7b89f10456d4057b124ad5302a](https://github.com/rackt/redux/commit/5ccf18b0e1055e7b89f10456d4057b124ad5302a)

**Comment by gaearon** [comment]
I'd imagine something like this: (pardon the renames)

```js
import { getVisibleTodos } from '../selectors'

const mapStateToProps = (state) => {
  return {
    todos: getVisibleTodos(state)
  }
}
...
const VisibleTodoList = connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoList)
```
I don't see the value in *selector suffix.

**Comment by jakegardner** [comment]
I agree terse names are better, but then it would be the same as the non-memoized version. What do you suggest?

Maybe I'm missing something. The output of createSelector() is a function that returns the memoized state. In the previous version of this document, that output was mapped as props to the child components of App. Now that we're doing the filtering in VisibleTodoList without a render method, I'm not sure how to map the output of createSelector() (via visibleTodosSelector) to child props.

**Comment by gaearon** [comment]
Doesn't my example show how? Call selector from mapStateToProps. I think :-)

**Comment by jakegardner** [comment]
Ok, I was trying to not modify the selector definition. But as you wish :-)

Have a look: [https://github.com/rackt/redux/pull/1295](https://github.com/rackt/redux/pull/1295)

**Comment by urbanvikingr** [comment]
Found a bug https://github.com/rackt/redux/pull/1332

**Comment by urbanvikingr** [comment]
@gaearon I'm updating `todomvc` example first. Already updated the source code, just need to complete tests. Check https://github.com/rackt/redux/pull/1334 for progress.

**Comment by urbanvikingr** [comment]
@gaearon I suggest that I rename `todomvc` example to `todos`. When I update the `todos-with-undo` example, I will use the file structure as in the `todos` example. Agree?

**Comment by gaearon** [comment]
The problem is that todomvc has more "stuff" than todos-with-undo. But I agree it's weird they are so different. Do as you think is best, and I'll take a look later!

**Comment by urbanvikingr** [comment]
OK, I'm going to rename `todomvc` to `todos` and fix the tests. Then we can decide whether to add additional functionality or not. Right now, I think that we need to align examples with documentation, so newcomers can follow.

**Comment by urbanvikingr** [comment]
Decided to keep the `todomvc` example and add `todos`. I could use some help with unit test of components and containers in the `todos` example.

**Comment by gaearon** [comment]
@urbanvikingr Where can I find the latest version you were working on?

**Comment by urbanvikingr** [comment]
bug in doc https://github.com/rackt/redux/pull/1332
todos https://github.com/rackt/redux/pull/1334 (just saw that it now has merge conflicts)

Perhaps we can merge doc updates?

**Comment by gaearon** [comment]
@urbanvikingr I’ll take care of merge conflicts, thanks. Let me take some time and I’ll merge `todos` in this PR.

**Comment by urbanvikingr** [comment]
@gaearon Awesome. Thanks. Then `todos` are merged, I will update doc references from `todomvc` to `todos`. Perhaps this weekend, I have time to look into unit testing of components and containers.

**Comment by gaearon** [comment]
Thanks to everyone involved! This is up: http://redux.js.org/docs/basics/UsageWithReact.html

**Pull request merged**